devices/dstt.cpp

```diff
@@ -187,72 +173,83 @@ class DSTT : Flashcart {
                 return true;
-        
+
         return false;
     }
 
     void Erase_Block(uint32_t offset, uint32_t length)
     {
-        dstt_flash_command(0x87, 0x5555, 0xAA);
-        dstt_flash_command(0x87, 0x2AAA, 0x55);
-        dstt_flash_command(0x87, 0x5555, 0x80);
-        dstt_flash_command(0x87, 0x5555, 0xAA);
-        dstt_flash_command(0x87, 0x2AAA, 0x55);
+        if (m_cmd_type == 1) {
+            dstt_flash_command(0x87, 0x5555, 0xAA);
+            dstt_flash_command(0x87, 0x2AAA, 0x55);
+            dstt_flash_command(0x87, 0x5555, 0x80);
+            dstt_flash_command(0x87, 0x5555, 0xAA);
+            dstt_flash_command(0x87, 0x2AAA, 0x55);
 
-        dstt_flash_command(0x87, offset, 0x30);
+            dstt_flash_command(0x87, offset, 0x30);
+        } else if (m_cmd_type == 2) {
+            dstt_flash_command(0x87, offset, 0x50); // Clear Status Register
+            dstt_flash_command(0x87, offset, 0x20); // Erase Setup
+            dstt_flash_command(0x87, offset, 0xD0); // Erase Confirm
+
+            // TODO: Timeout if something goes wrong.
+            while (!(dstt_flash_command(0, offset & 0xFFFFFFFC, 0) & 0x80));
+
+            dstt_flash_command(0x87, offset, 0x50); // Clear Status Register
+            // dstt_flash_command(0x87, offset, 0xFF); // Reset
+        }
 
         uint32_t end_offset = offset + length;
-        while (offset < end_offset)
+        for (; offset < end_offset; offset += 4)
         {
+            // TODO: Timeout if something goes wrong.
             while (dstt_flash_command(0, offset, 0) != 0xFFFFFFFF);
-            offset += 4;
         }
 
         dstt_reset();
     }
 
-    void Erase_Block_Type2(uint32_t offset, uint32_t length)
-    {
-        dstt_flash_command(0x87, offset, 0x50); // Clear Status Register
-        dstt_flash_command(0x87, offset, 0x20); // Block Erase 1
-        dstt_flash_command(0x87, offset, 0xD0); // Block Erase 2
-        
-        while (!(dstt_flash_command(0, offset & 0xFFFFFFFC, 0) & 0x80));
+    void readWord(uint32_t address, uint8_t *buffer) {
+        uint32_t data = dstt_flash_command(0, address, 0);
 
-        dstt_flash_command(0x87, offset, 0x50); // Clear Status Register
-        dstt_flash_command(0x87, offset, 0xFF); // Reset        
+        *buffer++ = (uint8_t)((data >> 0) & 0xFF);
+        *buffer++ = (uint8_t)((data >> 8) & 0xFF);
+        *buffer++ = (uint8_t)((data >> 16) & 0xFF);
+        *buffer++ = (uint8_t)((data >> 24) & 0xFF);
+    }
 
-        uint32_t end_offset = offset + length;
-        while (offset < end_offset)
-        {
-            while (dstt_flash_command(0, offset, 0) != 0xFFFFFFFF);
-            offset += 4;
-        }
+    uint32_t rawRead(uint32_t address, uint32_t length, uint8_t *buffer) {
+        return read_wrapper<DSTT, &DSTT::readWord, 4>(address, length, buffer);
     }
 
-    void Erase_Chip()
-    {
+    uint32_t rawErase(uint32_t address) {
```

Comment:
I can't even grasp how this works with the preprocessing? What's the aim of this function? Does it erase all blocks or? :S

Reply:
Ah I get that now, okay yeah that's much much better. Nice.